include/dlaf/eigensolver/tridiag_solver/rot.h

```diff
@@ -0,0 +1,175 @@
+  const SizeType mb = dist.blockSize().rows();
+
+  const SizeType sub_square_edge = (i_last + 1 - i_begin) * mb;
+  const GlobalElementSize sub_square_size(sub_square_edge, sub_square_edge);
+  const matrix::Distribution dist_sub(sub_square_size, dist.blockSize(), dist.commGridSize(),
+                                      dist.rankIndex(), dist.rankGlobalTile({i_begin, i_begin}));
+
+  // TODO check workspace distribution (partial allocation and its usage with dist_sub)
+  const matrix::Distribution dist_ws({dist.size().rows(), 1}, dist.blockSize(), dist.commGridSize(),
+                                     dist.rankIndex(), dist.sourceRankIndex());
+  // TODO allocate just sub-range
+  matrix::Panel<Coord::Col, T, D> workspace(dist_ws);
+
+  auto givens_rots_fn = [comm_row, dist_sub, mb](std::vector<GivensRotation<T>> rots,
+                                                 std::vector<matrix::Tile<T, D>> tiles,
+                                                 matrix::Tile<T, D> tile_ws) {
+    const SizeType m = dist_sub.localSize().rows();
+
+    auto getColPtr = [&dist_sub, &tiles, &tile_ws](const SizeType col_index, const bool hasIt) -> T* {
+      // TODO document column layout assumption
+      // TODO assume column layout and operate on the full column? also workspace must obey
+      if (hasIt) {
+        const LocalTileIndex tile_col{dist_sub.nextLocalTileFromGlobalElement<Coord::Row>(0),
+                                      dist_sub.nextLocalTileFromGlobalElement<Coord::Col>(col_index)};
+        const std::size_t linear_tile_col = to_sizet(dist_sub.localTileLinearIndex(tile_col));
+        return tiles[linear_tile_col].ptr(dist_sub.tileElementIndex({0, col_index}));
+      }
+      return tile_ws.ptr({0, 0});
+    };
+
+    ex::unique_any_sender<> serializer = ex::just();
+    for (const GivensRotation<T>& rot : rots) {
+      const SizeType j_x = rot.i / mb;
+      const SizeType j_y = rot.j / mb;
+
+      const comm::IndexT_MPI rankColX = dist_sub.template rankGlobalTile<Coord::Col>(j_x);
+      const comm::IndexT_MPI rankColY = dist_sub.template rankGlobalTile<Coord::Col>(j_y);
+
+      const bool hasX = dist_sub.rankIndex().col() == rankColX;
+      const bool hasY = dist_sub.rankIndex().col() == rankColY;
+
+      if (!hasX && !hasY)
+        continue;
+
+      T* col_x = getColPtr(rot.i, hasX);
+      T* col_y = getColPtr(rot.j, hasY);
+
+      const bool hasBothXY = hasX && hasY;
+
+      std::vector<ex::unique_any_sender<>> cps;
+      if (!hasBothXY) {
+        // TODO compute TAG
+
+        const comm::IndexT_MPI rank_partner = hasX ? rankColY : rankColX;
+
+        // TODO possible optimization, check if it is zero or not
+
+        const T* col_send = hasX ? col_x : col_y;
+        T* col_recv = hasX ? col_y : col_x;
+
+        cps.emplace_back(wrapper::scheduleSendCol<D, T>(comm_row, rank_partner, 0, col_send, m));
+        cps.emplace_back(wrapper::scheduleRecvCol<D, T>(comm_row, rank_partner, 0, col_recv, m));
```

Comment:
Here you can use `rot.i` and `rot.j` as tags depending on which one is sent and which one is received in the send and receive calls respectively.

Reply:
Exactly, since it is serialised it is not a problem...but! I should start using a tag because this can be called many times from the same caller which might pass the same communicator (i.e. without cloning it).

About this, in first instance I would add to the API a way to specify the tag to use for the call, leaving the responsibility to the caller to use independent tags for different calls to this.

I would prefer this solution because deciding tags independency from within it seems to me that we are embedding logic about its usage, which instead should be up to the caller.

What do you think? Would you prefer a different solution?

Reply:
> Exactly, since it is serialised it is not a problem...but! I should start using a tag because this can be called many times from the same caller which might pass the same communicator (i.e. without cloning it).

This is a good point!

> About this, in first instance I would add to the API a way to specify the tag to use for the call, leaving the responsibility to the caller to use independent tags for different calls to this.

Yes, I think this is reasonable!